spec/factories/middleware_datasource.rb

```diff
@@ -12,5 +12,6 @@
           :parent => :hawkular_middleware_datasource do
     name 'ExampleDS'
     nativeid 'Local~/subsystem=datasources/data-source=ExampleDS'
+    middleware_server { FactoryGirl.create(:middleware_server) }
```

Comment:
Why specifically is it in this factory and not the parent if the general model specifies a delegation that potentially breaks things?

Reply:
I did it this way, because the factory is called `hawkular_middleware_datasource_initialized`. That means, you want it to be ready to use. However, the parent one would assume you want a custom datasource, which you define on your own. And since for example all operations done on middleware datasource are done via mw server, you should have it already defined in the test context as well.

Reply:
>  I did it this way, because the factory is called hawkular_middleware_datasource_initialized. That means, you want it to be ready to use. However, the parent one would assume you want a custom datasource, which you define on your own.

This misses the point of inheritance and overriding in factories; FactoryGirl is specifically intended to allow you to have a fully working object (it has everything it needs to function) at the top level and customize as needed without redefining an entirely new factory. Don't bother changing it at this point, but for your own information...

`:hawkular_middleware_datasource_initialized` shouldn't exist, and the `:hawkular_middleware_datasource` should just have `middleware_server`, etc. If you want a custom datasource, you *do* define it on your own, and don't need a separate factory for it:

```ruby
# Give me a hawkular datasource that works, complete with the server:
FactoryGirl.create(:hawkular_middleware_datasource)

# Give me a hawkular datasource but I want to specify a custom server:
FactoryGirl.create(:hawkular_middleware_datasource, :middleware_server => my_custom_middleware_server, :nativeid => "blah")
```

Reply:
Sorry, I'm not understanding.  Are we saying that `hawkular_middleware_datasource` is still broken in some situations because we're not specifying a required middleware_server?

Reply:
I'd even say maybe `:middleware_datasource` is the broken one, since even `MiddlewareDatasource` `belongs_to :middleware_server`.

That said, all other uses of `hawkular_middleware_datasource` do provide either `middleware_server` or `server_id`, except for:

```
manageiq-ui-classic/spec/helpers/application_helper_spec.rb
1301:        FactoryGirl.create(:hawkular_middleware_datasource, :ext_management_system => ems, :name => "Test Middleware")
```

Reply:
(as for `:middleware_datasource`, there's one use which provides `middleware_server`, and 3 which don't:

```
manageiq-api/spec/requests/middleware_datasources_spec.rb
2:  let(:datasource) { FactoryGirl.create(:middleware_datasource) }

manageiq-api/spec/requests/collections_spec.rb
152:      FactoryGirl.create(:middleware_datasource)
480:      FactoryGirl.create(:middleware_datasource)
```
)

Reply:
@jrafanie, no, we have currently no other testcase which would be broken by this. But generally speaking yes, it can be used in "broken" way. But we have many factories which allows that unfortunately (the closest one is `:middleware_deployment` for example).

@chrisarcand, I was also surprised by existence of the `:hawkular_middleware_datasource_initialized` however in order to fix the Travis I did the minimally invasive thing around. Sure I'm aware you should define own values, and honestly it was my first attempt to fix this in https://github.com/ManageIQ/manageiq-ui-classic/pull/3044. And by request it was changed to the factory change...

I assume the best approach would be to change the top level `:middleware_datasource`  to include a [association](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#associations) to `:middleware_server`, is that correct?

Reply:
If we want to talk very objectively about how it should be, the top middleware datasource should provide everything needed for a middleware datasource to function, yes; and `:hawkular` should be a [FactoryGirl/Bot trait](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#traits). That last part I hadn't mentioned because we regrettably do this 'redefining a new factory instead of using traits' all over the codebase :)

> however in order to fix the Travis I did the minimally invasive thing around.

Yup, I respect that you're not the one who introduced this `_initialized` thing and we're venturing out of scope 😃 I would have asked you to change things around had I the opportunity to review this (Good morning from the American Central timezone, btw), but as I said, don't bother now that it's already merged and fixes your specific issue.

Reply:
@chrisarcand, Honestly, I do appreciate your review! And of course it would be nice to have it done properly. I'm gonna take a look into it and update at least the middleware factories to truly provide what they should. 😉

Reply:
@tumido 👍 Yes, doing a tactical fix in order to get travis green is most important especially if the "right" fix requires some discussion.  You did the right thing.  A followup PR to fix the parent factory is fine now that travis can has been fixed.